grpc/src/main/java/io/stargate/grpc/service/MessageHandler.java

```diff
@@ -104,30 +108,94 @@ protected MessageHandler(
     this.preparedCache = preparedCache;
     this.persistence = persistence;
     this.responseObserver = responseObserver;
+    this.retryPolicy = new DefaultRetryPolicy();
   }
 
   void handle() {
     try {
       validate();
+      executeWithRetry(0);
 
-      CompletionStage<ResultAndIdempotencyInfo> resultFuture =
-          prepare(false).thenCompose(this::executePrepared);
-      handleUnprepared(resultFuture)
-          .thenApply(this::buildResponse)
-          .thenCompose(this::executeTracingQueryIfNeeded)
-          .whenComplete(
-              (response, error) -> {
-                if (error != null) {
-                  handleException(error);
-                } else {
-                  setSuccess(response);
-                }
-              });
     } catch (Throwable t) {
       handleException(t);
     }
   }
 
+  private void executeWithRetry(int retryCount) {
+    executeQuery()
+        .whenComplete(
+            (response, error) -> {
+              if (error != null) {
+                RetryDecision decision = shouldRetry(error, retryCount);
+                if (decision == RetryDecision.RETRY) {
+                  executeWithRetry(retryCount + 1);
+                } else if (decision == RetryDecision.RETHROW) {
+                  handleException(error);
+                }
```

Comment:
This can be written as a switch statement.
And I think we throw on the default case, just in case another constant gets added in the future.

Reply:
done